pkg/cloud/aws/actuators/machine/actuator.go

```diff
@@ -167,32 +119,110 @@ func (a *Actuator) Create(ctx context.Context, cluster *clusterv1.Cluster, machi
 	}
 
 	controlPlaneMachines := GetControlPlaneMachines(clusterMachines)
+	if len(controlPlaneMachines) == 0 {
+		return errors.New("unable to proceed with machine creation until there is at least 1 control plane machine")
+	}
 
-	isNodeJoin, err := a.isNodeJoin(scope, controlPlaneMachines)
-	if err != nil {
-		return errors.Wrapf(err, "failed to determine whether machine %q should join cluster %q", machine.Name, cluster.Name)
+	/*
+	If new machine is a control plane machine, we may be in the following states:
+
+	1) We have not yet attempted to create any control plane VMs
+	2) We have created our first control plane VM, but it's not ready yet (apiserver still booting)
+	3) We have created at least one control plane VM and it's ready
+
+	To avoid race conditions, we use a single ConfigMap to coordinate.
+
+	1) Absence of ConfigMap: if the current machine is a control plane machine, try to create the ConfigMap with that machine's name in the
+	   data.
+			If creation succeeds, proceed to create the VM for the chosen machine.
+			If creation fails because it already exists, move on to checking if the control plane is ready.
+
+	2) Presence of ConfigMap: the chosen control plane machine should either be booting or ready. Try to reach the apiserver.
+			If we can reach the apiserver, we can proceed with a join (can be a control plane or regular node)
+			If we can't reach the apiserver, we need to keep waiting.
+	 */
+
+
+	// Except for the very first control plane machine, everything will be a join, so let's assume we're joining.
+	isNodeJoin := true
+
+	// TODO make sure this name isn't longer than the max allowed length
+	controlPlaneSyncConfigMapName := fmt.Sprintf("%s-controlplane", cluster.Name)
+
+	controlPlaneConfigMap, err := a.coreClient.ConfigMaps(cluster.Namespace).Get(controlPlaneSyncConfigMapName, metav1.GetOptions{})
+	if apierrors.IsNotFound(err) {
+		if machine.Spec.Versions.ControlPlane == "" {
+			// This isn't a control plane machine - have to wait
+			return errors.New("machine is not a control plane machine and control plane is not operational yet - need to wait")
+		}
+
+		blockOwnerDeletion := true
+		controlPlaneConfigMap = &apicorev1.ConfigMap{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: cluster.Namespace,
+				Name: controlPlaneSyncConfigMapName,
+				OwnerReferences: []metav1.OwnerReference{
+					{
+						APIVersion:	cluster.APIVersion,
+						Kind: cluster.Kind,
+						Name: cluster.Name,
+						UID: cluster.UID,
+						BlockOwnerDeletion: &blockOwnerDeletion,
+					},
+				},
+			},
+			Data: map[string]string{
+				"firstMachine": machine.Name,
+			},
+		}
+
+		_, err = a.coreClient.ConfigMaps(cluster.Namespace).Create(controlPlaneConfigMap)
+		if err != nil && !apierrors.IsAlreadyExists(err) {
```

Comment:
I'm not sure, but I think a conflict indicates that we lost the race, which suggests to me we want to move into the "wait for API availability" mode?

Reply:
The error returns a retry-able error to the controller and this machine will be reconciled in the next reconciliation loop in which the condition on L153 will be false and we'll wait for the api server to be available.
@ncdc, correct me if I am wrong 😊

Reply:
Oh, neat! I think we just want to remove the `&& !apierrors.IsAlreadyExists(err)` half of this condition then (or maybe return a different, more specific error in that case).

Reply:
Yes, I do think that would work